### ISSUE TITLE
Invalid characters in the Local File System cause rclone copy and sync to re-download existing files under the new name

### ISSUE THREAD

**Issue by Aleytia** [rclone/rclone#2633]
#### What is the problem you are having with rclone?
In Google Drive, you can store folders under any name. This implies that local file systems (Windows, in this case) may not be able to perfectly copy file names, since they contain invalid file names. After rclone replaces these invalid characters, it does not seem to compare files under the new (fixed) path under the other configuration destination, causing rclone to re-download all files into the destination.

For example, if I have a folder named "TEST:TEST" in Google Drive, and run copy/sync, rclone will download it to the Windows FS as "TEST_TEST", and then copy all files inside (remote) "TEST:TEST" into (local) "TEST_TEST". Running the same copy/sync command, however, then causes rclone to re-download all files in "TEST:"TEST to "TEST_TEST" again, as it does not seem to check the correct folder name's contents.

#### What is your rclone version (output from `rclone version`)
rclone v1.43.1, go version: go1.11

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Windows 10 Pro, 64-bit

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive to Windows FS

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
rclone copy (remote:) (local_path)

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

PS C:\Users\<USERNAME>\Documents\rclone> .\rclone.exe -vv copy <GOOGLE_DRIVE_CONFIG> <LOCAL_PATH>
2018/10/08 14:05:18 DEBUG : rclone: Version "v1.43.1" starting with parameters ["C:\\Users\\<USERNAME>\\Documents\\rclone\\rclone.exe" "-vv" "copy" "<GOOGLE_DRIVE_CONFIG>" "<LOCAL_PATH"]
2018/10/08 14:05:18 DEBUG : Using config file from "C:\\Users\\<USERNAME>\\.config\\rclone\\rclone.conf"
2018/10/08 14:05:24 INFO  : Local file system at \\?\<LOCAL_PATH>: Waiting for checks to finish
2018/10/08 14:05:24 INFO  : Local file system at \\?\<LOCAL_PATH>: Waiting for transfers to finish
2018/10/08 14:05:24 NOTICE: Local file system at \\?\<LOCAL_PATH>: Replacing invalid characters in "\\<LOCAL_PATH>\\TEST:TEST\\<FILE NAME>" to "\\\\?\\E:\\<LOCAL_PATH>\\TEST_TEST\\<FILE NAME>"
2018/10/08 14:06:23 INFO  :
Transferred:      740.996M / 2.140 GBytes, 34%, 11.344 MBytes/s, ETA 2m7s
Errors:                 0
Checks:                 7 / 7, 100%
Transferred:            0 / 1, 0%
Elapsed time:      1m5.3s
Transferring:
 *   ...<FILE NAME>: 33% /2.140G, 12.537M/s, 1m55s

2018/10/08 14:07:23 INFO  :
Transferred:        1.465G / 2.140 GBytes, 68%, 11.969 MBytes/s, ETA 57s
Errors:                 0
Checks:                 7 / 7, 100%
Transferred:            0 / 1, 0%
Elapsed time:      2m5.3s
Transferring:
 *   ...<FILE NAME>: 68% /2.140G, 12.283M/s, 56s

2018/10/08 14:08:19 INFO  : <FILE NAME>: Copied (new)
2018/10/08 14:08:19 INFO  :
Transferred:        2.140G / 2.140 GBytes, 100%, 12.093 MBytes/s, ETA 0s
Errors:                 0
Checks:                 7 / 7, 100%
Transferred:            1 / 1, 100%
Elapsed time:      3m1.2s

2018/10/08 14:08:19 DEBUG : 4 go routines active
2018/10/08 14:08:19 DEBUG : rclone: Version "v1.43.1" finishing with parameters ["C:\\Users\\<USERNAME>\\Documents\\rclone\\rclone.exe" "-vv" "copy" "GOOGLE_DRIVE:" "LOCAL_PATH"]

**Comment by B4dM4n** [comment]
I can see what is happening there. When `sync` is listing the drive folder, it is seeing the `TEST:TEST` folder, but when listing the local folder it only sees `TEST_TEST`. Therefore `TEST:TEST` is assumed missing and will be transferred every time.

I'm not sure how to solve this with the current `sync` implementation. Maybe remotes which can alter incoming file names should implement a optional interface like `NameMapper` which can be used to change file and folder names into the destination version.

**Comment by ncw** [comment]
This is an old problem! I think the original scheme of replacing characters with underscores is flawed as it loses information.

Maybe we should make the local backend when running Windows do the same sort of name translations that the OneDrive backend does (to and from Unicode equivalents) which are much more likely to preserve information.

The local backed is the only backed which does that lossy transform. All the others try quite hard to preserve all characters.

It might be annoying for users if we switched the encoding though...

Your idea of name mapper would work too, it could be used in the march package where similar translations are applied at the moment (eg case folding and Unicode normalisation)

**Comment by B4dM4n** [comment]
> Maybe we should make the local backend when running Windows do the same sort of name translations that the OneDrive backend does (to and from Unicode equivalents) which are much more likely to preserve information.

If the OneDrive remote already does this, using it for the local filesystem sounds like a good idea. Encountering the replacement chars on a real remote is pretty unlikely.

> It might be annoying for users if we switched the encoding though...

You mean from replacing to `_` to the Unicode equivalents? This would lead to duplicate folders when using copy, but after one run, the names would be stable and the repeated copy as reported in this issue would be resolved.
A simple test shows the current behavior broken and has the potential to loose data with only the notice `Replacing invalid characters`:
```
C:\Users\fmoeller\Desktop>rclone copy -vv TestDrive:TestInvalid TestInvalid
2018/10/10 12:27:30 DEBUG : rclone: Version "v1.43-DEV" starting with parameters ["rclone" "copy" "-vv" "TestDrive:TestInvalid" "TestInvalid"]
2018/10/10 12:27:30 DEBUG : Using config file from "C:\\Users\\fmoeller\\.config\\rclone\\rclone.conf"
2018/10/10 12:27:31 INFO  : Local file system at \\?\C:\Users\fmoeller\Desktop\TestInvalid: Waiting for checks to finish
2018/10/10 12:27:31 INFO  : Local file system at \\?\C:\Users\fmoeller\Desktop\TestInvalid: Waiting for transfers to finish
2018/10/10 12:27:31 NOTICE: Local file system at \\?\C:\Users\fmoeller\Desktop\TestInvalid: Replacing invalid characters in "\\Users\\fmoeller\\Desktop\\TestInvalid\\a?a\\file.txt" to "\\\\?\\C:\\Users\\fmoeller\\Desktop\\TestInvalid\\a_a\\file.txt"
2018/10/10 12:27:31 INFO  : a?a/file.txt: Copied (new)
2018/10/10 12:27:31 NOTICE: Local file system at \\?\C:\Users\fmoeller\Desktop\TestInvalid: Replacing invalid characters in "\\Users\\fmoeller\\Desktop\\TestInvalid\\a:a\\file.txt" to "\\\\?\\C:\\Users\\fmoeller\\Desktop\\TestInvalid\\a_a\\file.txt"
2018/10/10 12:27:31 INFO  : a:a/file.txt: Copied (new)
2018/10/10 12:27:31 INFO  :
Transferred:            12 / 12 Bytes, 100%, 6 Bytes/s, ETA 0s
Errors:                 0
Checks:                 0 / 0, -
Transferred:            2 / 2, 100%
Elapsed time:        1.7s

2018/10/10 12:27:31 DEBUG : 6 go routines active
2018/10/10 12:27:31 DEBUG : rclone: Version "v1.43-DEV" finishing with parameters ["rclone" "copy" "-vv" "TestDrive:TestInvalid" "TestInvalid"]


C:\Users\fmoeller\Desktop> rclone tree TestDrive:TestInvalid
/
├── a:a
│   └── file.txt
└── a?a
    └── file.txt

2 directories, 2 files

C:\Users\fmoeller\Desktop> rclone tree TestInvalid
/
└── a_a
    └── file.txt

1 directories, 1 files
```

Using the Unicode equivalents is probably the best way handle invalid characters and keep the folder structure of the source intact.

A bonus feature would be a `--remove-unicode-equivalents` flag, which forces rclone to replace the Unicode equivalents in the source back. This would allow the following flow without creating duplicates: `Drive -> Windows Local -> External HDD -> Linux Local -> Drive`

> Your idea of name mapper would work too, it could be used in the march package where similar translations are applied at the moment (eg case folding and Unicode normalisation)

Originally I had this idea in mind to allow file name changes during document conversions in `drive`, eg `file.doc -> file.docx`.
I scrapped it first because implementing it in the `march` package seemed to hard for a simple file rename, but after seeing the implications of not handling colliding names, the `march` package should handle this even with replacement chars implemented for the local remote.

### ToDo ###

1. implement Unicode equivalents for the local remote to reduce the chance of data loss and fixing the repoted issue
2. handle colliding names in the `march`package with a `NameMapper` interface
3. allow the forced removal of Unicode equivalents for remotes that usually don't use them with a `--remove-unicode-equivalents` flag

**Comment by ncw** [comment]
> ### ToDo
> 1. implement Unicode equivalents for the local remote to reduce the chance of data loss and fixing the repoted issue

I'm quite keen on this idea.  It would simplify the tests greatly too as they jump through loads of hoops just for the lossy windows local backend (see `fstest/fstest.go` `Item.Path` and `Item.WinPath`)

There are 3 versions of the name replacement code already

It would be nice to unify them into a sensible library (in "lib") and make them parameter driven since they all do roughly the same thing just with a few tweaks.

```
$ find . -name replace.go
./backend/jottacloud/replace.go
./backend/opendrive/replace.go
./backend/onedrive/replace.go
```

> 2. handle colliding names in the `march`package with a `NameMapper` interface

This idea is orthogonal to the other two ideas and would be useful for those pesky drive docs!

> 3. allow the forced removal of Unicode equivalents for remotes that usually don't use them with a `--remove-unicode-equivalents` flag

That sounds like a good idea too.

It would be great if you wanted to have a go at any or all of these :-)

**Comment by B4dM4n** [comment]
I did some testing to figure out the invalid characters and needed replacements for some remotes. It turned out that there are many variations on the supported characters.
I created a [Google Spreadsheet](https://docs.google.com/spreadsheets/d/1fHxVJL2Y6ITjPgsz6YmA8VaBjhsb6KhEDWiKoF3tcaU/edit?usp=sharing) containing my findings.

There a not many surprises, but I found two characters in OpenDrive that could not be created and downloaded correctly: `&` and `+`. I will create a separate issue to keep track of this.

Another question is how to handle the control characters `0x01-0x31` in addition to the invalid utf8 strings. Some remotes support them properly and others not. I suggest to create a `fserrors.ErrorInvalidPathName` so that creating a file or folder with invalid characters (not replaceable) returns a unified error on the destination side.
This is probably a more consistent way to handle these unusual names instead of silently applying a lossy conversion.

**Comment by ncw** [comment]
> I did some testing to figure out the invalid characters and needed replacements for some remotes. It turned out that there are many variations on the supported characters.

Looks like you discovered `rclone info` :-)

> I created a [Google Spreadsheet](https://docs.google.com/spreadsheets/d/1fHxVJL2Y6ITjPgsz6YmA8VaBjhsb6KhEDWiKoF3tcaU/edit?usp=sharing) containing my findings.

Great!

What are the `xbf-�-` and `xfe-�-` lines - invalid unicode?

> _Edit:_
> These results are obtained with all existing replacing disabled (but with restore enabled as seen in the ＼ line )

I note that google drive file names can have `/` in (which I'm sure you know but it isn't in your 'sheet).
 
> There a not many surprises, but I found two characters in OpenDrive that could not be created and downloaded correctly: `&` and `+`. I will create a separate issue to keep track of this.

Now fixed :-)
 
> Another question is how to handle the control characters `0x01-0x31` in addition to the invalid utf8 strings. Some remotes support them properly and others not. I suggest to create a `fserrors.ErrorInvalidPathName` so that creating a file or folder with invalid characters (not replaceable) returns a unified error on the destination side.
> This is probably a more consistent way to handle these unusual names instead of silently applying a lossy conversion.

I've got some quite extensive notes about this problem which I'll paste below to see if it gives you any ideas.  It gives a scheme for what to do with the control characters.  It also has the idea of a loss-less transform.  The idea isn't complete though as I haven't worked through exactly where the transform and inverse should be applied.

I note also that we could easy adjust the matching in March to match % with ％ (unicode equivalent) which would go a long way towards ironing out problems with a potentially lossy transform.

Massive info dump follows ... ;-)

Problem
-------

We need the file name translations done in rclone to be reversible otherwise syncing does not work properly.

The windows stuff make the tests a nightmare...

Unicode normalization & case insensitivity causes syncing problems
  => multiple ways of representing the same filename

Invariant
---------

Each remote can round trip any sequence of UTF-8 characters, including `/` (including a name like `http://example.com`)

FIXME once implemented will need a no-escape option for Put so we can make rclone info work again

Escaping
--------

Use a standard escaping mechanism using

To go from ASCII range to FULLWIDTH range add 0xFE80

>>> for i in range(32,128): print chr(i), unichr(i+0xfee0)

>>> print u"".join(unichr(i) for i in range(33,127))
!"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~

>>> print u"".join(unichr(i+0xfee0) for i in range(33,127))
！＂＃＄％＆＇（）＊＋，－．／０１２３４５６７８９：；＜＝＞？＠ＡＢＣＤＥＦＧＨＩＪＫＬＭＮＯＰＱＲＳＴＵＶＷＸＹＺ［＼］＾＿｀ａｂｃｄｅｆｇｈｉｊｋｌｍｎｏｐｑｒｓｔｕｖｗｘｙｚ｛｜｝～

https://en.wikipedia.org/wiki/Unicode_control_characters#Control_pictures

Add 0x2400 to ASCII control characters 0-32 to get their picture
>>> print u"".join(unichr(i+0x2400) for i in range(0,33))
␀␁␂␃␄␅␆␇␈␉␊␋␌␍␎␏␐␑␒␓␔␕␖␗␘␙␚␛␜␝␞␟␠

Note that 127 DEL is seperate

>>> print unichr(0x2421)
␡

Need an encode which can escape or unescape any of those 128 character positions.

We also need a scheme to use when the original has the replacement character in.

Propose using another unicode character as an escaping mechanism - maybe ‛ 0x201B - SINGLE HIGH REVERSED 9 QUOTATION MARK 

So ‛X would decode to X and ‛‛ would decode to ‛

Quoting needs to encode any possible encoded character
  - Just choose the punctuation out of the oblique list which gives some forward compatibility
  - ！＂＃＄％＆＇（）＊＋，－．／：；＜＝＞？＠［＼］＾＿｀｛｜｝～
  - ␀␁␂␃␄␅␆␇␈␉␊␋␌␍␎␏␐␑␒␓␔␕␖␗␘␙␚␛␜␝␞␟␠
  - ␡
  - ‛

```
So encode looks like
  for each character
    if on forbidden list
      substitute
    elif on quotabe list
      quote
```

```
Decode looks like
  if quote
    read next and put it literrally into stream
    if literal wasn't quotable then put them both in
  elif on forbidden translated list
    substitute
```

We MUST ONLY encode characters which mustn't exist on the local disk
otherwise this leads to problems, eg let's say we did encode / for
google drive.  Since people put / in the names of things in google
drive say we have a file called "a / b" this encodes to "a ／ b" on
local.  This would then decode back to "a ／ b" which would cause a
duplicated file.

Encodings
---------

Need different encodings for different parts

core encoding - used in the core and for all remotes unless otherwise stated
  - encodes / and 0-31,127
  - Can use this to fix objects with names like "http://example.com" and "a_dir/" #1492

windows local encoding
  - encodes < > : " / \ | ? * 0-31,127 etc (does it allow 127 or not?)
  - Folders can't end with spaces

unix local encoding
  - encodes / and 0 only

drive encoding
  - encodes 0-31, 127 (not /)

onedrive encoding
  - encode "/"  "\"  "*"  "<"  ">"  "?"  ":"  "|"  "#"  "%"
  - encode 0-31, 127
  - Folder names can't end with a period '.'
  - OneDrive for Business file or folder names cannot begin with a tilde '~'
  - Apparently file names can't start with space either

NEED to know exactly which remotes can take which ASCII characters from 0-127.  Perhaps write a program to do that? (`rclone info`)

ACD can have /r on - do we leave that?
S3 can have all unicode characters apparently...


Code
----

Make Encoding with functions

// Encodes raw go string into encoding
func (e *Encoding) Encode(raw string) (encoded string)

// Decodes encoded input string into raw go string
func (e *Encoding) Decode(encoded string) (raw string)

Will need to do a lot of transcoding...

// remote names to disk
remoteEncoding.Encode(Standard.Decode(remote))

// disk to remote name
remoteEncoding.Decode(Standard.Encode(remote))

All we actually need is

func (e *Encoding) ToStandard(encoded string) (standard string)
func (e *Encoding) FromStandard(standard string) (encoded string)

We could potentially do that as we go along.

How do we do that nicely in the code without stuffing each remote full
of translate, untranslate...

  - could say each remote uses standard encoding except for
    - windows local
    - unix local
    - onedrive
    - drive
      - this could use a much simplified version just translating ／ to / (if not escaped)
  - this isn't going to work for the escaping of existing quotable characters though
    

Scenario
--------

drive
  drive encoding "*/␉"
  raw encoding "*/\t"
  unix local "*／\t"
  windows local "＊／␉"
  onedrive encoding as "＊／␉"
  and back to drive as "*/␉"

drive
  drive encoding: "/␍／"
  raw: "/\n／"
  standard encoding: "／␍‛／"
  raw: "/\n／"
  local encoding: "／\n‛／"


Windows
=======

From http://www.secfs.net/winfsp/develop/sshfs/

The reserved Windows characters are:

<   >   :   "   /   \   |   ?   *
any character between 0 and 31

POSIX only has two reserved characters: slash (/) and NUL.

So how do we map between the two? Luckily this problem has been solved
before by "Services for Macintosh" (SFM), "Services for UNIX" (SFU)
and Gygwin. The solution involves the use of the Unicode "private use
area". When mapping a POSIX path to Windows, if we encounter any of
the Windows reserved characters we simply map it to the Unicode range
U+F000 - U+F0FF. The reverse mapping from Windows to POSIX is obvious.


From the forum
===============

https://forum.rclone.org/t/another-special-character-issue-this-time-the-lovely-and-characters/306/9

Is there any chance of an option to treat one end as the source of truth and do escaping and unescaping when syncing files with the other end?

e.g. On Google Drive: “foo/bar”
…syncs locally to: “foo%2fbar”

My worry is that renaming invalid symbols to Unicode characters loses information — what if I have files using those characters? (And I do.) I won’t be able to restore the backup and end up with an exact copy of what I had, because rclone can’t tell which Solidus code points need to be turned back into slashes and which ones don’t.

So have a perfectly reversible scheme

**Comment by Rhilip** [comment]
Same problem find when I use rclone to copy/sync files , which named with FULLWIDTH character , from `Google Drive` to `OneDrive`. And those FULLWIDTH character will make files transfer again and again in my cron program.
I hope This problem can be solved ASAP.

**Comment by B4dM4n** [comment]
> Looks like you discovered rclone info :-)

Yes, `rclone info`was very helpful :smiley: 
I had to extend it a bit to also test characters at the left and right edges of file names.

> What are the xbf-�- and xfe-�- lines - invalid unicode?

Those are invalid UTF-8 bytes, which can not appear on their own. But testing those is still a bit unreliable, since I'm not sure if `rclone ls` outputs them as is.
The tests showed that most of the backends don't support invalid UTF-8 bytes. All backends using `encoding/json` and `encoding/xml` to encode a string field in a struct will replace those bytes with the error rune �.
The local backend on Windows has a similar issue. Since the Windows API uses UTF-16, Go has to convert it's UTF-8 strings, which also replaces invalid bytes with the error rune ([UTF16FromString](https://github.com/golang/go/blob/master/src/syscall/syscall_windows.go#L44) converts the `string` to `[]rune` and `rune` can't represent those invalid bytes).

> I note that google drive file names can have / in (which I'm sure you know but it isn't in your 'sheet).

Sure I know :wink: 
`rclone info` skips those and therefor the sheet ignores them.

> Massive info dump follows ... ;-)

Thanks for the detailed information, it was very helpful.

The implementation of a `lib/encoder` package based on your description should be done now: [master...B4dM4n:unicode-replacer](https://github.com/ncw/rclone/compare/master...B4dM4n:unicode-replacer)

It contains a `MultiEncoder` struct, which can be configured to escape and quote sets of characters. 
The `Encode` method only replaces characters based on the given configuration and quotes only characters that are a result of such a replacement and the `Decode` method will only undo the changes from `Encode`.
The package also provides a `Standard` encoder, which uses the following configuration: `EncodeZero | EncodeSlash | EncodeCtl | EncodeDel`
This should be the encoding that all paths and names use for input, output and between remotes.

There is currently one issue with encoding \ under windows. `fspath.Parse` and the local backend use `filepath.ToSlash` to convert all \ to /. This makes it impossible to use \ on Windows for all backends.
Should we add `EncodeBackSlash` to the `Standard` encoder? This would allow ＼ to be used on Windows and ＼ or \ on all other systems (as `Decode` wont change the \ ).

**Comment by B4dM4n** [comment]
> > Those are invalid UTF-8 bytes, which can not appear on their own. But testing those is still a bit unreliable, since I'm not sure if `rclone ls` outputs them as is.
> 
> Hmm, I would have thought it would, but getting them into `ls` would be difficult!

It might work fine, but I had trouble copying those bytes around. I'm just not sure how accurate my test scripts work when processing the output of `rclone ls` with these bytes. Therefor those lines in the spreadsheet might not always  be correct.

> Hmm, yes until this moment rclone has assumed
> 
> * all backends are UTF-8 only
> * no backends support invalid UTF-8
> * the local backend replaces all invalid UTF-8 with a warning
> 
> This hasn't caused too many problems, except for Linux users with non-UTF8 filenames. The [local backend docs](https://rclone.org/local/#filenames) encourage them to use `convmv` to convert the character set of the files.
> 
> We should probably continue issuing the warning as some users will prefer to convmv their files and have the files look correct in the dropbox website (for example). Some users will just want the files to sync correctly though and not want to change the encoding.

Good point, I did't think about encodings like latin1 when handling invalid UTF-8 bytes.
So in addition of converting them to the quoted format there will be a warning as there is currently?

> > The local backend on Windows has a similar issue. Since the Windows API uses UTF-16, Go has to convert it's UTF-8 strings, which also replaces invalid bytes with the error rune ([UTF16FromString](https://github.com/golang/go/blob/master/src/syscall/syscall_windows.go#L44) converts the `string` to `[]rune` and `rune` can't represent those invalid bytes).
> 
> How much of a problem is this in practice? I don't think I've seen any bug reports from Windows users about this.

This has a similar impact for the local backend under Windows, as using the `encoding/json` package in the other backends.
Creating files with invalid UTF-8 bytes on Windows will silently convert those bytes to the error rune before the name goes into the Windows API.
This conversion also happens when reading a directory with the Windows API. If a filename contains a invalid UTF-16 word, this gets mapped to the error rune in Go.
I did not test this, but this mapping should allow Go to see those files, but prevent Go from accessing them as it can't sent the exact same UTF-16 sequence to the Windows API.

There is a similar issue for the [Rust](https://github.com/rust-lang/rust/issues/12056) language, but but I could not find one for Go.

Because of this, we should enable `EncodeInvalidUtf8` under Windows.
 
> Wow - a very impressive piece of work :-)
> 
> I like your autogenerated test suite :-)

Thanks a lot.
It was quite a bit of work, but it was worth it. The new encoder is about 8 to 10 time faster than the old method and does a lot less allocations.

The generated test helped me track down a few missed cases.
 
> So you are thinking of the case when there is a filename on a remote (drive say) which has a "" in and a Windows user wants to access it on the command line...
>
> That particular bit of code has caused problems in the past. Windows users are quite inclined to use `\` when they meant `/` in remotes (understandably) so the filepath.ToSlash was put in to help them with that.
> 
> I'm not sure making the user type `＼` on the command line to read a file with a `\` in is a workaround many users will ever discover, or figure out how to type `＼`!
> 
> Perhaps we should just change that one place to turn `＼` into `\` in Windows rather than change the `Standard` encoding?

Even if we change the `Standard` encoding to include `EncodeBackSlash`, Linux users would't need to enter ＼ on the command line. They could still use the regular \ and it would be used as is. But the output  of rclone would change from \ to ＼. Windows users have no other choice and need to enter ＼ on the command line if they wand to use \ as a name character in either case.

So we could make rclone behave nearly the same on Windows and Linux if we add `EncodeBackSlash` to the `Standard` encoding or we introduce a Windows specific decoding that is done after `fspath.Parse` and replaces `＼` with `\`:
```
root = encoder.MultiEncoderEncodeStandard | EncodeBackSlash).ToStandardPath(root)
```
This would make rclone behave asymmetrical when outputting a path like:
`C:\folder＼a\file.txt`
would be output as
`C:\folder\a\file.txt`

Another option would be to change the `Standard` encoding only for Windows to include `EncodeBackSlash`. This would have the benefit that Linux users could use and see \ as normal and Windows users need to use `＼` if they need to.
 
> Nice! That will even be reasonably visually readable as in ‛7F‛87. It is probably better than the macOS way in the info dump (mapping reserved characters to U+F000 - U+F0FF). The expansion may be important if the file names get too long. I think `convmv` is probably a better solution in the long run though!

Using `convmv` is definitely the better way to handle those filenames. This expansion should not occur often, otherwise someone is using really strange filenames and praying they get copied correctly (between multiple filesystems) . 

> Just a thought: mountlib or vfs should be doing translations on Windows too shouldn't it? At the moment we will be showing filenames with invalid Windows characters. The `Standard` encoding

Good point, I put it on the todo list. It shouldn't be to hard to implement.

> What do you want do next? More testing? Merge some stuff? Do you want to try to hit the v1.45 release (currently scheduled for 24th November)?

Most of the changes should be ready to be merged. Once we decide how to handle \ on Windows, the `lib/encoder` is ready. What is still missing and needs to be done is the user documentation.
If we get it done this week the v1.45 release should be fine.
When merging the backends, I suggest to do a PR for each one, since some are not straight forward or contain other fixes.

**Comment by gabriellluz** [comment]
As for OneDrive,  they will reject files with the following characters:
The name can not contain any of the following characters:
" Quotation mark

Number sign
% Percent Sign
& Ampersand

Asterisk
: Colon
< Less than Symbol
Greater than Symbol
? Question Mark
\ Backslash
{ Left Brace
} Right Brace
| Pipe
~ Tilde
/ Slash

The name cannot begin or end with a period
The name cannot contain a double period
The name cannot start or end with a space:
Bad " Game files", Good: "Game files"
Bad "Game files ", Good: "Game files"

**Comment by zephyrus00jp** [comment]
Just my two cents worth.

I recently disengage one of PCs from dropbox (Dropbox recently restricted the number of connected devices to three for free users. I had to disengage existing PC so that a new PC at the office could be connected.)

And instead I tried to see if I can use rlcone to copy modified files in a batch manner.
Under linux, it worked beautifully.

But under Windows, I got problems with the mentioned invalid characters in file paths (valid as linux path, but invalid as Windows).
I thought I did not have any issues when I used Dropbox and was wondering how Dropbox handled this.
Simple: Dropbox never bothered to copy files that would contain invalid characters on the destination file. (I think Dropbox central server is linux-based and has a much permissive character sets for path names than Windows.)

I suggest to introduce an operation mode to handle the case.
Create an option to suppress the copying of files that would have invalid path names.
rclone might want to issue a warning of files that are not copied due to this. But this warning should be
suppressed if the user so desires.
This operation mode matches the behavior of Dropbox and then
rclone will match my requirement perfectly :-)

I think substitution on the fly will inevitably  cause conflicts with existing filenames even, and so should be avoided although reasonably powerful substitution may be made available with such caveat clearly explained to the users.

Simply try to do nothing  (no copying) a la Dropbox is a valid solution and we should weigh the wisdom of a big service company in this regard.

I have been using Dropbox very conveniently and forgot about this "do not copy files with invalid paths" until today. I must have learned it in the early days of usage and unconsciously avoided creating such paths. It is just that saving an e-mail article as a file from within Thunderbird mail client creates a file
with a pathname that starts with its subject and also includes the address of the sender.
This means such files often have "Re: ..." at the beginning (note the COLON character) and
"<" ">" as in "<foobar@example.com>". "@" seems OK even under the windows. 

TIA

See discussion in
[1] https://www.dropboxforum.com/t5/Dropbox/Invalid-characters-in-filename-cause-undesired-behaviour-in/idi-p/203897

[2] a dropbox user can check files that have problematic characters using the following URL.
https://www.dropbox.com/bad_files_check

[3] Dropbox explains what are invalid characters under certain OS.
https://help.dropbox.com/syncing-uploads/files-not-syncing
